src/main/java/week3/WordSearch.java

```diff
@@ -0,0 +1,195 @@
+package week3;
+
+
+import java.util.ArrayList;
+import java.util.BitSet;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class WordSearch {
+
+  /*
+   * Given a grid of letters and a dictionary, find all the words from the dictionary that can be
+   * formed in the grid.The rules for forming a word:
+   *    ● You can start at any position.
+   *    ● You can move to one of the 8 adjacent cells (horizontally/vertically/diagonally).
+   *    ● You can't visit the same cell twice in the same word.Assume you have a dictionary class
+   *      with these two methods:
+   *    ● isWord(string): Returns whether the given string is a valid word.
+   *    ● isPrefix(string): Returns whether the given string is a prefix of at least one word in
+   *      the dictionary.
+   *
+   * Your function receives the number of rows, number of columns, a 2-dimensional array of
+   * characters (of the native char data type), and the dictionary. You should return the set of all
+   * words found.
+   */
+
+  private int rows;
+  private int cols;
+
+  /*
+   * Iteratively try to find longer words
+   * initializeMap = get all valid starting points on the grid (valid prefixes of length 1)
+   *
+   * Iteration: while there are longer words found, try to append the next character to the valid
+   * prefixes in the set.
+   *
+   * Time Complexity: Initialization + (number of iterations * cost of each iteration)
+   *                  O(|grid|*Words) + (O(|grid|) * O(|grid|*|grid| * |words|))
+    *                  O(|grid|^3*|words|)
+   *
+   *      where |grid| = rows*cols is the size of the grid and
+   *            |Words| = size of the dictionary / number of words in the dictionary
+   *
+   *  Intuitively, at worst the whole grid forms a word and thus we iterate over the while loop
+   *  |grid| times.
+   *
+   *  In each iteration, we can have all the grid being a prefix of some word, and the processing of
+   *  each path involves O(1) to get adjacent cells and O(|words|) to validate newPrefix
+   */
```

Comment:
I disagree with your complexity analysis. How many possible paths are there through the grid, and can you create an input that will try to evaluate them all?